File: tests/functional/m68k/test_icount_period.py
#!/usr/bin/env python3
#
# Exact icount period option tests
#
# Copyright 2026 Bryce Lanham
#
# This work is licensed under the terms of the GNU GPL, version 2 or later.
# See the COPYING file in the top-level directory.

from qemu_test import QemuSystemTest


class IcountPeriod(QemuSystemTest):

    def _launch_valid(self, period):
        self.vm.set_machine('none')
        self.vm.add_args('-S', '-nodefaults',
                         '-icount', f'ns-per-insn={period}')
        self.vm.launch()
        self.assertEqual(self.vm.cmd('query-status')['status'], 'prelaunch')

    def _launch_invalid(self, option, message):
        self.vm.add_args('-S', '-display', 'none', '-machine', 'none',
                         '-nodefaults', '-icount', option)
        self.vm.set_qmp_monitor(enabled=False)
        self.vm.launch()
        self.vm.wait()
        self.assertEqual(self.vm.exitcode(), 1)
        self.assertRegex(self.vm.get_log(), message)

    def test_exact_period(self):
        self._launch_valid(143)

    def test_maximum_period(self):
        self._launch_valid(1024)

    def test_zero_period(self):
        self._launch_invalid('ns-per-insn=0',
                             r'Invalid ns-per-insn value')

    def test_period_too_large(self):
        self._launch_invalid('ns-per-insn=1025',
                             r'Invalid ns-per-insn value')

    def test_period_conflicts_with_shift(self):
        self._launch_invalid('shift=7,ns-per-insn=143',
                             r'shift and ns-per-insn are mutually exclusive')


if __name__ == '__main__':
    QemuSystemTest.main()

File: tests/unit/test-icount-period.c
/*
 * Exact icount period tests
 *
 * Copyright 2026 Bryce Lanham
 *
 * This work is licensed under the terms of the GNU GPL, version 2 or later.
 * See the COPYING file in the top-level directory.
 */

#include "qemu/osdep.h"
#include "exec/icount.h"

static void test_period_range(void)
{
    g_assert_false(icount_period_valid(0));
    g_assert_true(icount_period_valid(1));
    g_assert_true(icount_period_valid(MAX_ICOUNT_PERIOD_NS));
    g_assert_false(icount_period_valid(MAX_ICOUNT_PERIOD_NS + 1ULL));
}

static void test_period_conversion(void)
{
    g_assert_cmpint(icount_period_to_ns(6984, 143), ==, 998712);
    g_assert_cmpint(icount_period_to_ns(6984, 128), ==, 893952);
}

static void test_period_rounding(void)
{
    const int64_t count = 6984;
    const int64_t exact = count * 143;

    g_assert_cmpint(icount_period_round(exact, 143), ==, count);
    g_assert_cmpint(icount_period_round(exact + 1, 143), ==, count + 1);
}

int main(int argc, char **argv)
{
    g_test_init(&argc, &argv, NULL);
    g_test_add_func("/icount/period/range", test_period_range);
    g_test_add_func("/icount/period/conversion", test_period_conversion);
    g_test_add_func("/icount/period/rounding", test_period_rounding);
    return g_test_run();
}
